governance/xc-admin/packages/xc-admin-frontend/components/tabs/UpdatePermissions.tsx

```diff
@@ -1,88 +1,364 @@
+import { proposeInstructions } from 'xc-admin-common'
+import { ClusterContext } from '../../contexts/ClusterContext'
 import { usePythContext } from '../../contexts/PythContext'
+import {
+  getMultisigCluster,
+  UPGRADE_MUTLTISIG,
+  useMultisig,
+} from '../../hooks/useMultisig'
 import CopyIcon from '../../images/icons/copy.inline.svg'
+import { capitalizeFirstLetter } from '../../utils/capitalizeFirstLetter'
 import ClusterSwitch from '../ClusterSwitch'
+import Modal from '../common/Modal'
+import EditButton from '../EditButton'
 import Loadbar from '../loaders/Loadbar'
 
 interface UpdatePermissionsProps {
   account: string
-  pubkey?: PublicKey
+  pubkey: string
+  newPubkey?: string
 }
 
-const UpdatePermissionsRow: React.FunctionComponent<UpdatePermissionsProps> = ({
-  account,
-  pubkey = new PublicKey(0),
-}) => {
-  return (
-    <tr key={account} className="border-t border-beige-300">
-      <td className="py-3 pl-4 pr-2 lg:pl-14">{account}</td>
-      <td className="py-3 pl-1 lg:pl-14">
-        <div
-          className="-ml-1 inline-flex cursor-pointer items-center px-1 hover:bg-dark hover:text-white active:bg-darkGray3"
-          onClick={() => {
-            copy(pubkey.toBase58())
-          }}
-        >
-          <span className="mr-2 hidden lg:block">{pubkey.toBase58()}</span>
-          <span className="mr-2 lg:hidden">
-            {pubkey.toBase58().slice(0, 6) +
-              '...' +
-              pubkey.toBase58().slice(-6)}
-          </span>{' '}
-          <CopyIcon className="shrink-0" />
-        </div>
-      </td>
-    </tr>
-  )
+const DEFAULT_DATA: UpdatePermissionsProps[] = [
+  {
+    account: 'Master Authority',
+    pubkey: new PublicKey(0).toBase58(),
+  },
+  {
+    account: 'Data Curation Authority',
+    pubkey: new PublicKey(0).toBase58(),
+  },
+  {
+    account: 'Security Authority',
+    pubkey: new PublicKey(0).toBase58(),
+  },
+]
+
+const BPF_UPGRADABLE_LOADER = new PublicKey(
+  'BPFLoaderUpgradeab1e11111111111111111111111'
+)
+
+const columnHelper = createColumnHelper<UpdatePermissionsProps>()
+
+const defaultColumns = [
+  columnHelper.accessor('account', {
+    cell: (info) => info.getValue(),
+    header: () => <span>Account</span>,
+  }),
+  columnHelper.accessor('pubkey', {
+    cell: (props) => {
+      const pubkey = props.getValue()
+      return (
+        <>
+          <div
+            className="-ml-1 inline-flex cursor-pointer items-center px-1 hover:bg-dark hover:text-white active:bg-darkGray3"
+            onClick={() => {
+              copy(pubkey)
+            }}
+          >
+            <span className="mr-2 hidden lg:block">{pubkey}</span>
+            <span className="mr-2 lg:hidden">
+              {pubkey.slice(0, 6) + '...' + pubkey.slice(-6)}
+            </span>{' '}
+            <CopyIcon className="shrink-0" />
+          </div>
+        </>
+      )
+    },
+    header: () => <span>Public Key</span>,
+  }),
+]
+
+type PubkeyChanges = {
```

Comment:
Maybe add some types here? There are only three possible keys "MasterAuthority", etc...

Reply:
You could make it a Record<"Master authority"| "Security authority" | "Data Curation", string >